### PR TITLE
fix(webkit): always push state changes to the provisional page

### PR DESCRIPTION
**Pull request by yury-s**
Copied from https://github.com/microsoft/playwright/pull/555

### CODE REVIEW

File: src/webkit/wkPage.ts [hunk 1]
```diff
@@ -79,8 +81,6 @@ export class WKPage implements PageDelegate {
     this.rawKeyboard.setSession(session);
     this._addSessionListeners();
     this._workers.setSession(session);
-    // New bootstrap scripts may have been added during provisional load, push them
-    // again to be on the safe side.
     if (this._bootstrapScripts.length)
```
Comment (dgozman) [inline src/webkit/wkPage.ts]: I think we can remove this.

File: src/webkit/wkPage.ts [hunk 2]
```diff
@@ -121,8 +121,10 @@ export class WKPage implements PageDelegate {
       promises.push(session.send('Page.overrideUserAgent', { value: contextOptions.userAgent }));
     if (this._page._state.mediaType || this._page._state.colorScheme)
       promises.push(WKPage._setEmulateMedia(session, this._page._state.mediaType, this._page._state.colorScheme));
-    if (isProvisional)
-      promises.push(this._setBootstrapScripts(session));
+    if (isProvisional && this._bootstrapScripts.length) {
```
Comment (dgozman) [inline src/webkit/wkPage.ts]: Can we drop isProvisional check here?

File: src/webkit/wkPage.ts [hunk 3]
```diff
@@ -139,6 +141,30 @@ export class WKPage implements PageDelegate {
...
+    this._provisionalPage = new WKProvisionalPage(provisionalSession, this);
+  }
+
+  onProvisionalLoadCommitted(session: WKSession) {
+    assert(this._provisionalPage);
+    assert(this._provisionalPage!._session === session);
+    this._provisionalPage!.commit();
+    this._provisionalPage!.dispose();
+    this._provisionalPage = null;
+    this.setSession(session);
```
Comment (dgozman) [inline src/webkit/wkPage.ts]: Can we make setSession private now?

File: src/webkit/wkPage.ts [hunk 4]
```diff
@@ -139,6 +141,30 @@ export class WKPage implements PageDelegate {
...
+  }
+
+  onSessionDestroyed(session: WKSession, crashed: boolean) {
+    if (this._provisionalPage && this._provisionalPage._session === session) {
+      this._provisionalPage.dispose();
+      this._provisionalPage = null;
+      return;
+    }
+    if (this._session === session && crashed)
+      this.didClose(crashed);
```
Comment (dgozman) [inline src/webkit/wkPage.ts]: Is this the only callsite of didClose?

**Review by dgozman — APPROVED** [review]

File: src/webkit/wkPage.ts (hunk 1 above)
Comment (yury-s) [reply src/webkit/wkPage.ts]: Of cause.

File: src/webkit/wkPage.ts (hunk 2 above)
Comment (yury-s) [reply src/webkit/wkPage.ts]: Yes.

File: src/webkit/wkPage.ts (hunk 3 above)
Comment (yury-s) [reply src/webkit/wkPage.ts]: Yes, done.

File: src/webkit/wkPage.ts (hunk 4 above)
Comment (yury-s) [reply src/webkit/wkPage.ts]: There is still one more in wkPageProxy.ts

**Pull request closed without merging**